noteable_magics/datasource_postprocessing.py

```diff
@@ -158,3 +163,58 @@ def postprocess_snowflake(
         db = dsn_dict['database']
 
         dsn_dict['database'] = f'{db}/{schema}'
+
+
+@register_postprocessor('sqlite')
+def postprocess_sqlite(
+    datasource_id: str, dsn_dict: Dict[str, str], create_engine_kwargs: Dict[str, Any]
+) -> None:
+    """Expect to have either a URL provided which we should download and place into TMPDIR as the database file,
+    or no such thing implying to run in memory mode.
+    """
+
+    if 'database' in dsn_dict:
+        cur_path = dsn_dict['database']
+        if cur_path == '' or cur_path == ':memory:':
+            # Empty path is alias for :memory:, and is fine.
+            return
+
+        # If it smells like a URL, we should download, stash it into a tmpfile,
+        # and respell dsn_dict['database'] to point to that file. Any exceptions in here
+        # will spoil the datasource.
+        parsed = urlparse(dsn_dict['database'])
+
+        # (Sigh, 'mock' as scheme due to cannot cleanly pytest requests-mock http or https urls
+        #  for reasons I trust from the requests-mocks docs)
+        if parsed.scheme in ('http', 'https', 'ftp', 'mock'):
+            max_download_time = 10  # seconds. What do we want this to be?
+
+            resp = requests.get(dsn_dict['database'], stream=True, timeout=max_download_time)
+
+            resp.raise_for_status()
+
+            # Save to a durable tmpfile
+            with NamedTemporaryFile(delete=False) as outf:
+                for chunk in resp.iter_content(chunk_size=2048):
+                    outf.write(chunk)
+
+            # Point to the resulting file.
+            dsn_dict['database'] = cur_path = outf.name
+
+        # The database file should resolve to somewhere /tmp-y (for now)
+        allowed_parents = ['/tmp']
+        if os.environ.get('TMPDIR'):
+            # And also TMPDIR, which might not be in /tmp.
+            #
+            # On OSX, /var is symlink to /private/var, so to get test suite passing
+            # need to canonicalize the path so the .startswith() test will work.
+            # (on OSX at least under pytest, the NamedTemporaryFile above will be
+            # something like /var/tmp/... , which is really /private/var/tmp/...)
+            allowed_parents.append(str(Path(os.environ.get('TMPDIR')).resolve()))
+
+        requested = str(Path(cur_path).resolve())
+
+        if not any(requested.startswith(allowed_parent) for allowed_parent in allowed_parents):
```

Comment:
`.is_relative_to` on the `Path` class is a bit more reliable than `startswith`

Reply:
👍